hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java

```diff
@@ -163,7 +163,6 @@ boolean isDone() {
     }
 
     boolean shouldSchedule() {
-        // Stashed tasks that never got scheduled, and weren't cancelled in-between
         return future == null && this.resultRef.get() == null;
```

Comment:
Should the future field not be volatile? I'm not fully understanding the thread interaction of the ScheduledTaskDescriptor but it seems to be touched by multiple threads but I'm failing to see a happens before relation on the future field.

Reply:
Yes, there is sth fishy here. I want to believe (I also remember) that, I had a reason for not keeping them volatile, but I seem unable to remind myself of it. I will check the changes on the file since it was refactored a bit, but as it is now, I believe these two fields need to become volatile. Good catch!

Reply:
@pveentjer 
Ok, I believe I managed to reconstruct my memory, on the choice of not using volatile on those fields. Regardless of the finding though, it's worth documenting the interactions here, since it's not obvious at first sight, not even to the author (me :/ )

So we have two flows in the Service.

A. Flow for partitioned tasks, tasks that got vanilla scheduled, or scheduled on particular key-owner (partition). Scheduling and bookkeeping happen on partition threads, every interaction with the offloaded runner thread, goes through atomic ops, e.g. State. Result, Stats

B. Flow for non-partitioned tasks, tasks that get scheduled for a specific member. Interaction with these happen through generic threads, so they need special care. The scheduling part, which is responsible for creating the entry, scheduling the runnable, assigning the future, is done under a lock for that Member (see https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorMemberOwnedContainer.java) and the result is stored in a ConcurrentMap. All further reads on that task, get the task from the Map. The future never changes, its only not-final due to the serialization needs for the partitioned flavor of this feature. it's important to remind at this stage, that Member owned tasks, don't support any durability options, no backups, no migrations whatsoever.

I will document the above in the class to make it more visible, and also help me remember faster next time. The correctness of the above part, aside, due to your poking, I was able to discover a breach of this flow, and that is here  https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/TaskRunner.java#L64 the descriptor is shared with the runner thread, and we are accessing the type of the task there. This should be happening in the constructor of this class, which make me think whether I should be making these fields volatile regardless ... 

I will fix the latter asap.